br/pkg/lightning/restore/data_sample_check.go

```diff
@@ -0,0 +1,263 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package restore
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+	"unicode/utf8"
+
+	"github.com/pingcap/errors"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/tidb/br/pkg/lightning/config"
+	"github.com/pingcap/tidb/br/pkg/lightning/log"
+	"github.com/pingcap/tidb/br/pkg/lightning/mydump"
+	"github.com/pingcap/tidb/br/pkg/utils"
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/types"
+)
+
+type dataSampleCheck struct {
+	controller *Controller
+
+	checkCfg      *config.CheckOnly
+	checkTemplate Template
+	wg            sync.WaitGroup
+
+	totalRows                    atomic.Int64
+	totalColumnCountMismatchRows atomic.Int64
+	totalInvalidCharRows         atomic.Int64
+}
+
+func newDataSampleCheck(controller *Controller) *dataSampleCheck {
+	return &dataSampleCheck{
+		controller:    controller,
+		checkTemplate: NewSimpleTemplate(),
+		checkCfg:      controller.cfg.CheckOnly,
+	}
+}
+
+func (d *dataSampleCheck) logCheckFailedRow(row []types.Datum, fileMeta mydump.SourceFileMeta, offset int64, errMsg string) {
+	var sb strings.Builder
+	for i, d := range row {
+		if i > 0 {
+			sb.WriteString(", ")
+		}
+		// see Datum.String()
+		v := d.GetValue()
+		if b, ok := v.([]byte); ok && d.Kind() == types.KindBytes {
+			v = string(b)
+		}
+		sb.WriteString(fmt.Sprintf("%v", v))
+	}
+	log.L().Error("data check failed", zap.String("raw-data", sb.String()),
+		zap.String("file", fileMeta.Path),
+		zap.Int64("offset", offset),
+		zap.String("error", errMsg))
+}
+
+func (d *dataSampleCheck) checkRoutine(ctx context.Context, fileChan chan *sampledDataFileInfo, errChan chan error) {
+	defer d.wg.Done()
+
+	var resultErr error
+	defer func() {
+		errChan <- errors.Trace(resultErr)
+	}()
+
+	rc := d.controller
+	for fileInfo := range fileChan {
+		fileMeta := fileInfo.DataFile.FileMeta
+		fileParser, err := rc.createDataFileParser(ctx, fileMeta, config.DefaultCSVDataInvalidCharReplace)
+		if err != nil {
+			if errors.Cause(err) != context.Canceled {
+				resultErr = err
+			}
+			return
+		}
+
+		var rowsChecked, columnCountMismatchRows, invalidCharRows, lastOffset int64
+		columnCount := len(fileInfo.TableInfo.Columns)
+		var i int64
+		for {
+			i++
+			if d.checkCfg.Rows != config.CheckAllRows && i > d.checkCfg.Rows {
+				break
+			}
+			err = fileParser.ReadRow()
+			if err != nil {
+				if errors.Cause(err) == io.EOF {
+					break
+				}
+
+				if errors.Cause(err) != context.Canceled {
+					resultErr = err
+				}
+				fileParser.Close()
+				return
+			}
+
+			row := fileParser.LastRow()
+			rowsChecked++
+			if len(row.Row) != columnCount {
```

Comment:
1. what if table schema contains generated columns?
2. lightning support auto generate default value for missing field if the field has default value

Reply:
3. How this column check work with `ignoreColumns` feature?

Reply:
> How this column check work with ignoreColumns feature?

Reply:
Does the current logic can alow import following data?
table schema:
````
create table t (val varchar(16), id int primary key auto_increment);
```
input csv, `test.t.0.csv`:
```
test1
test2
```

Reply:
no. without header, columns in csv should match ddl

Reply:
?